### PR TITLE
Preallocate nodes in advance of their need to avoid the allocation being behind a mutex.

### PR DESCRIPTION
**Pull request by Tilps**

### CODE REVIEW

**Comment by Tilps** [comment]
I think this is faster, but its probably less than 5% improvement, so its not proving easy gather data trivially showing a large improvement like my prior PR.

**Comment by borg323** [comment]
Some tests with the 4 threads multiplexing the random backend show this gains about 3%.
The following figure shows all the results (10 runs for each) sorted in ascending order
image

File: src/mcts/node.h [hunk 1]
```diff
@@ -131,6 +131,13 @@ class Node {
   // Takes pointer to a parent node and own index in a parent.
   Node(Node* parent, uint16_t index) : parent_(parent), index_(index) {}
 
+  // Performs construction time type initialization. For use only with a node
+  // that has not been used beyond its construction.
+  void Reinit(Node* parent, uint16_t index) {
```
[the remark this answers is not in the document]
Comment (Tilps) [reply src/mcts/node.h]: Done.

**Comment by Tilps** [comment]
Thanks - those results seem to match up with what I was seeing by eyeballing multiple runs...

**Pull request merged**